core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java

```diff
@@ -517,6 +513,40 @@ boolean isExpiration() {
          return false;
       }
 
+      private void onSuccess() {
+         synchronized (this) {
+            // send succeed.
+            if (log.isTraceEnabled()) {
+               log.tracef("[IRAC] State.onSuccess for key %s (status=%s)", key, stateStatus);
+            }
+            if (stateStatus != StateStatus.SENDING) {
+               // discarded or overwritten by another write operation
+               // do not send the cleanup command.
+               return;
+            }
+            stateStatus = StateStatus.COMPLETED;
+         }
+         onRoundCompleted(IracResponseCollector.Result.OK, false);
+         removeStateFromCluster(this);
+      }
```

Comment:
This isn't exactly what I had in mind... My point was that I hate it code that does similar things should live in methods with similar names to let the user know that they sit at the same abstraction level.

You already have methods `discard()` and `sendFail()`, which only update the `stateStatus` and log, so I was suggesting

```suggestion
      synchronized void sendSuccess() {
         // send succeed.
         if (log.isTraceEnabled()) {
            log.tracef("[IRAC] State.onSuccess for key %s (status=%s)", key, stateStatus);
         }
         if (stateStatus != StateStatus.SENDING) {
            // discarded or overwritten by another write operation
            // do not send the cleanup command.
            return;
         }
         stateStatus = StateStatus.COMPLETED;
      }
```

The two lines at the end of `onSuccess()` would remain in the `onResponse()` method.

Reply:
Yeah, I missed that!
It still bothers me that `onRoundCompleted()` is called both from onResponse and from onResponse->onSuccess, but I can live with it ;)